src/herder/Upgrades.cpp

```diff
@@ -88,6 +126,53 @@ Upgrades::UpgradeParameters::toJson() const
     return out.str();
 }
 
+std::string
+Upgrades::UpgradeParameters::toDebugJson() const
+{
+    Json::Value upgradesJson;
+    Json::Reader reader;
+    reader.parse(toJson(), upgradesJson);
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    if (mConfigUpgradeSet != nullptr)
+    {
+        // combine the key and actual config upgrade set under a single Json
+        // value
+        Json::Value configUpgradeSetJson;
+        Json::Reader reader;
+        reader.parse(mConfigUpgradeSet->toJson(), configUpgradeSetJson);
+        upgradesJson["configupgradeinfo"]["configupgradeset"] =
+            configUpgradeSetJson;
+        upgradesJson["configupgradeinfo"]["configupgradesetkey"] =
+            upgradesJson["configupgradesetkey"];
+        upgradesJson.removeMember("configupgradesetkey");
+    }
+#endif
+    Json::StyledWriter writer;
+    return writer.write(upgradesJson);
+}
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+void
+Upgrades::UpgradeParameters::setConfigUpgrades(stellar::AbstractLedgerTxn& ltx,
+                                               ConfigUpgradeSetKey const& key)
+{
+    mConfigUpgradeSet = ConfigUpgradeSetFrame::makeFromKey(ltx, key);
+}
```

Comment:
we should probably check that it's valid here as well, that way if bad keys are referenced, commandHandler will fail properly

Reply:
I explicitly call `getConfigUpgradeSet` in the command handler to make sure something was set.

Reply:
This is something different.
you can point to a malformed `ConfigUpgradeSet`, in which case returning "not null" is still not enough:
* not sorted properly (`isValidXDR` set to false)
* contains bad settings `isValidForApply` returns false

Reply:
Ah good point. It's safe to throw an exception here right? This function is only called in the command handler and in the Herder on startup. The alternative is to call `mConfigUpgradeSet.reset()`;

Reply:
I think this is irrelevant now